### ISSUE TITLE
Being able to get separate IServiceCollection instance per each branch though IApplicationBuilder.MapWhen

### ISSUE THREAD

**Issue by tugberkugurlu** [aspnet/HttpAbstractions#492]
When I call `IApplicationBuilder.MapWhen`, I am able to create a different pipeline but all my pipelines rely on the same `IServiceCollection` instance. It makes sense for [`ApplicationServices`](https://github.com/aspnet/HttpAbstractions/blob/c80946260a17f5ad046dc77728f5c291ddc235d4/src/Microsoft.AspNet.Http/ApplicationBuilder.cs#L35-L45) but doesn't for [`RequestServices`](https://github.com/aspnet/HttpAbstractions/blob/67aa2546a87f0617005e6f5571ed7fed8ab2f088/src/Microsoft.AspNet.Http/Features/IServiceProvidersFeature.cs#L10).

I want to be able to provide different `IServiceCollection` instances per branch. Defaulting to use the same one wouldn't be bad.

**Comment by tugberkugurlu** [comment]
I am not sure if this is the correct repository for this issue though.

**Comment by khellang** [comment]
> It makes sense for `ApplicationServices` but doesn't for `RequestServices`.

Why not? These are not different collections, they represent the same services in different scopes.

> I want to be able to provide different `IServiceCollection` instances per branch.

Why? Is this some kine of multi-tenancy requirement? You want different implementations of services for each branch? How would you wire those up? You have a single method to wire up your services; `ConfigureServices`. After that, the `IServiceProvider` is immutable.

**Comment by tugberkugurlu** [comment]
> Why not? These are not different collections, they represent the same services in different scopes.

Because I don't think you would want different implementations of `IApplicationEnvironment`, `IHostingEnvironment` for each pipeline. When you get to the `Startup` class, ApplicationServices are all decided for you.

> Why? Is this some kine of multi-tenancy requirement?

No, when you are creating a different pipeline, you want to run in an isolated state from the other pipelines. This is especially important when the middleware depend on services (see https://github.com/aspnet/Security/issues/591) and you want to use the same middleware in two pipelines. 

> You want different implementations of services for each branch?

Not always, it should be up to application developer to decide this when wiring up the pipelines.

> How would you wire those up? You have a single method to wire up your services; ConfigureServices. After that, the IServiceProvider is immutable.

I don't know, that's just an implementation detail (e.g. have named scopes for each pipeline and be able to register services into named scopes, etc.). At the end of the day, I want to be able to run pipelines in isolation.

**Comment by brockallen** [comment]
IIRC, there's an open issue in the DI repo asking for a feature called "multi-tenancy" (which makes no sense to me as a name). But it's basically asking for DI to be different at different places in the pipeline (which I'm +1 for supporting). With this feature you could actually host multiple instances of MVC in the pipeline. As of now, you can't really given the single shared DI config.

**Comment by Tratcher** [comment]
You could do this with the [RequestServicesContainerMiddleware] (https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerMiddleware.cs), adding it at the start of each branch. You can also replace the IApplicationBuilder.ApplicationServices at the start of each branch (or should there be an overload of Map* that allows you to rebuild services for that branch?).

**Comment by tugberkugurlu** [comment]
> (or should there be an overload of Map* that allows you to rebuild services for that branch?)

That would be good. It would basically abstracts the logic away which is nice to a certain extent. This would not cover the below scenario I assume though:

 - I want to register `ServiceA` as application wide singleton.
 - I want to register `ServiceB` as singleton (would end up being singleton per isolation (branch)).
 - I want to register `ServiceC` as scoped (would end up being scoped per isolation (branch)).

**Comment by brockallen** [comment]
> You could do this with the RequestServicesContainerMiddleware,

Is this the officially sanctioned way in ASP.NET for the "multi-tenancy DI"? Or is/was there other work that was planned or needed to really make this work?

**Comment by Tratcher** [comment]
I'm not aware of any planned/official answer, I'm just brainstorming how you might accomplish it.

**Comment by muratg** [comment]
@sebastienros Were you able to achieve this in Orchard impl?  Any plans to document it for other folks?

cc: @Eilon

**Comment by csunali** [comment]
A scenario where this might be useful is; imagine a `MapWhen` call for `/Backend` Uri where we register MVC services with some custom options and we also have the default branch of services where we register MVC services again.

Not sure if my explanation was clear so I will try to give an example:

I would like the backend to have custom routing conventions:

```
            // Add MVC services to the services container.
            services.AddMvc(options =>
            {
                options.Conventions.Add(new CustomBackendRoutingConvention());
            });
```

However, I don't want the default branch to have that custom routing convention so I can just:

```
            // Add MVC services to the services container.
            services.AddMvc();
```

**Comment by muratg** [comment]
Orchard vNext was able to achieve some of the goals here. Feel free to check out https://github.com/OrchardCMS/Orchard2

**Comment by sebastienros** [comment]
I will need some time to write some relatively complete explanation on how we managed to do it. In a nutshell, to be able to support different services per tenants we need a specific scope for each tenant. We did it without using any external IOC framework but it would be easier with one. 

First, what we did is inject the `ServiceCollection` that will be used to configure the container:
https://github.com/OrchardCMS/Orchard2/blob/2b697c183afb6f08409dd66d4b5e63ad46c4b5ab/src/Orchard.Web/Startup.cs#L20

With this we are able to reflect on the list of services, and redefine them. When the site is started, we create a new service collection by translating all the singleton scoped services to tenant singletons, by creating a new scope for each tenant, and creating new registrations. We also inject the host instances to each tenant scope.
https://github.com/OrchardCMS/Orchard2/blob/bab799d3d461ecbd9126986f2d6b1aad4bb4781b/src/Orchard.DependencyInjection/ServiceProviderExtensions.cs

All these service providers are saved globally, then as @Tratcher mentioned we have our custom middleware that will set the `RequestServices` collection for the current scope one:
https://github.com/OrchardCMS/Orchard2/blob/2e0b9e22af792fe23970fb621aec06beb06ea623/src/Orchard.Hosting.Web/OrchardContainerMiddleware.cs#L55

What we are achieving is a third level of service provider (Application, Tenant, Request) that is necessary to support per-tenant custom services. This way we can either have different implementations for each tenant, or also different service lists. In Orchard's case it's important as each tenant can have a different set of modules/services enabled.

There are other parts of the system that are not explained here but are important. Routing is one, as each tenant needs different routes, but also each route triggers the current tenant. And then there is tenant lifetime management. That would be a great Build talk I assume (in case my managers get the message).
